### PR TITLE
boot: zephyr: serial_recovery: Add boot mode enter ability

### PR DESCRIPTION
**Pull request by nordicjm**
Adds an optional entrance method for mcuboot's serial recovery by using Zephyr's boot mode data retention system, this allows for an application to set the retained data and reboot into the bootloader.

### CODE REVIEW

**Review by de-nordic — APPROVED** [review]
Looks OK. Tested with https://github.com/zephyrproject-rtos/zephyr/pull/55125 and seems to work fine.

**Comment by nordicjm** [comment]
@d3zd3z @utzig can you review please?

File: boot/zephyr/main.c
Comment (nvlsianpu) [inline boot/zephyr/main.c:1]: I think function for enter serial recovery can be extracted:
``` C
#ifdef CONFIG_MCUBOOT_INDICATION_LED
        gpio_pin_set_dt(&led0, 1);
#endif

        mcuboot_status_change(MCUBOOT_STATUS_SERIAL_DFU_ENTERED);
        rc = boot_console_init();
        bootmode_clear();
        boot_serial_start(&boot_funcs);
```

For instance bootmode_clear(); might be transferred to it as a callback to be called.

File: boot/zephyr/Kconfig.serial_recovery [hunk 1]
```diff
@@ -142,8 +133,26 @@ config BOOT_SERIAL_ENCRYPT_EC256
...
 	  described under "ECIES-P256 encryption" in docs/encrypted_images.md.
 
-config BOOT_SERIAL_WAIT_FOR_DFU
-	bool "Wait for a prescribed duration to see if DFU is invoked by receiving a mcumgr comand"
+menu "Entrance methods"
+
+menuconfig BOOT_SERIAL_ENTRANCE_GPIO
+	bool "GPIO"
+	default y
+	depends on GPIO
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply boot/zephyr/Kconfig.serial_recovery:141]: If the board selects it by default or has a driver that enables it by default, yes, which should be all boards as per the zephyr porting guide.

File: boot/zephyr/main.c
Comment (nordicjm) [reply boot/zephyr/main.c:1]: Updated

**Review by nvlsianpu — APPROVED** [review]

File: boot/zephyr/Kconfig.serial_recovery [hunk 2]
```diff
@@ -155,6 +164,29 @@ config BOOT_SERIAL_WAIT_FOR_DFU_TIMEOUT
...
-	  timeout in ms for MCUboot to wait to allow for DFU to be invoked.
+	  Timeout in ms for MCUboot to wait to allow for DFU to be invoked.
+
+config BOOT_SERIAL_BOOT_MODE
+	bool "Check boot mode via retention subsystem"
+	depends on RETENTION_BOOT_MODE
+	help
+	  Allows for entering serial recovery mode by using Zephyr's boot mode
+	  retention system (i.e. an application must set the boot mode to stay
+	  in serial recovery mode and reboot the module).
```
Comment (wbober) [inline boot/zephyr/Kconfig.serial_recovery:175]: What is module?

Comment (nordicjm) [reply boot/zephyr/Kconfig.serial_recovery:175]: As in what is the retention system? Or another question?

Comment (wbober) [reply boot/zephyr/Kconfig.serial_recovery:175]: I was referring to module in `... reboot the module`. How about `device` or `SoC`?

**Review by d3zd3z — COMMENTED** [review]
Looks ok, other than my question about it being a choice.

File: boot/zephyr/Kconfig.serial_recovery [hunk 3]
```diff
@@ -142,8 +133,26 @@ config BOOT_SERIAL_ENCRYPT_EC256
 	  encryption mechanism used in this case is ECIES using primitives
 	  described under "ECIES-P256 encryption" in docs/encrypted_images.md.
 
-config BOOT_SERIAL_WAIT_FOR_DFU
-	bool "Wait for a prescribed duration to see if DFU is invoked by receiving a mcumgr comand"
+menu "Entrance methods"
+
+menuconfig BOOT_SERIAL_ENTRANCE_GPIO
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply boot/zephyr/Kconfig.serial_recovery:138]: Multiple modes can be enabled, e.g. GPIO will be enabled by default but would also probably make sense to enable GPIO and "if application does not exist" - could even then change it so GPIO is only enabled by default if the line is present in the dts but these would all be future plans

**Comment by nordicjm** [comment]
@davidvincze I assume you're now helping with mcuboot? If so, any chance you could review/approve/merge this PR?

**Review by d3zd3z — APPROVED** [review]

**Pull request merged**